source/validate_builtins.cpp

```diff
@@ -0,0 +1,1474 @@
+#include <string>
+
+#include "diagnostic.h"
+#include "opcode.h"
+#include "val/validation_state.h"
+
+using libspirv::Decoration;
+using libspirv::DiagnosticStream;
+using libspirv::Instruction;
+using libspirv::ValidationState_t;
+
+namespace { // utils
+
+bool isIncludedIn(uint32_t elem, std::vector<uint32_t> vec) {
+  return std::find(vec.begin(),vec.end(),elem) != vec.end();  
+}
+
+// Returns whether the given instruction has a BuiltIn decoration
+bool isBuiltIn(Instruction inst, ValidationState_t& vstate) {
+  const auto& decorations = vstate.id_decorations(inst.id());
+  return std::any_of(
+      decorations.begin(), decorations.end(), [](const Decoration& d) {
+        return SpvDecorationBuiltIn == d.dec_type();
+      });
+}
+
+// Returns the SpvBuiltIn_enum of the first BuiltIn decoration of a instruction
+uint32_t getBuiltInEnum(Instruction inst, ValidationState_t& vstate) {
+  const auto& decorations = vstate.id_decorations(inst.id());
+  auto it = std::find_if(
+      decorations.begin(), decorations.end(), [](const Decoration& d) {
+        return SpvDecorationBuiltIn == d.dec_type();
+      });
+  assert(it != decorations.end());
+  return it->params()[0];
+}
+
+// Returns the Storage class of the instruction
+uint32_t getStorageClass(Instruction inst) {
+  switch (inst.opcode()) {
+    case SpvOpTypePointer: return inst.words()[2];
+    case SpvOpTypeForwardPointer: return inst.words()[2];
+    case SpvOpVariable: return inst.words()[3];
+    case SpvOpGenericCastToPtrExplicit: return inst.words()[4];
+    default: return SpvStorageClassMax;
+  }
+}
+
+// TODO(jcaraban): how to find the entry_points in a more direct way?
+// Returns the execution models of the entry points linked to the builtin
+std::vector<uint32_t> getExecutionModels(
+      ValidationState_t& vstate, Instruction inst) {
+  std::vector<uint32_t> exec_vec;
+  // Walks the entry functions in the module
+  for (auto entry_func : vstate.entry_points()) {
+    const auto &interfaces = vstate.entry_point_interfaces(entry_func);
+    // Filtering those interfacing the built-in of interest
+    if (isIncludedIn(inst.id(),interfaces)) {
+      auto func_inst = vstate.FindDef(entry_func);
+      assert(func_inst);
+      // The entry function uses link to the entry points
+      for (auto pair : func_inst->uses()) {
+        auto entry_point = pair.first;
+        if (entry_point->opcode() == SpvOpEntryPoint) {
+          // The entry points contain the execution model
+          auto exec_model = entry_point->words()[1];
+          exec_vec.push_back(exec_model);
+        }
+      }
+    }
+  }
+  return exec_vec;
+}
+
+// Ensures the excution model is included in the list of compatible models
+spv_result_t checkExecutionModel(ValidationState_t& vstate,
+      uint32_t exec_model, std::vector<uint32_t> models) {
+  if (not isIncludedIn(exec_model,models)) {
+    return vstate.diag(SPV_ERROR_INVALID_ID)
+        << "Built-in variables are restricted to only certain"
```

Comment:
It's much more friendly for the error message to state the builtin and the execution model that it was found in.

Reply:
Agree, it was in my plans. I will update the tests accordingly.